### PR TITLE
[geometry/optimization] Enable rounding of GCS solution when solving convex relaxation of ShortestPath

### PR DESCRIPTION
**Pull request by mpetersen94**
Adds an optional rounding step to the SolveShortestPath. Currently off by default to preserve the current behaviors.

Depends on #17396 

+@RussTedrake for feature review
cc @TobiaMarcucci

- - -
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17796)

### CODE REVIEW

**Review by mpetersen94 — COMMENTED** [review]
This currently only returns the result of the rounding, which means we can't access the convex relaxation result to check the gap without running the solver again.  Any advice on whether this matters/how to change the call signature to enable getting both results back would be helpful!

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17796#-:-NAkDTIf7xKMmlZtporK:byq9psd)* status: LGTM missing from assignee RussTedrake(platform), needs at least two assigned reviewers, missing label for release notes (waiting on @RussTedrake)

**Review by RussTedrake — COMMENTED** [review]
need tests here, too..

> Reviewed 2 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17796#-:-NBEXJZx3PC3-XIZkjKm:bazjy0c)* status: 3 unresolved discussions, LGTM missing from assignee RussTedrake(platform), needs at least two assigned reviewers, missing label for release notes (waiting on @mpetersen94 and @RussTedrake)

___
*[`geometry/optimization/graph_of_convex_sets.h` line 34 at r1](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NAkYrVm-j0vvL0ET683:-NAkYrVm-j0vvL0ET684:bwg2i54) ([raw file](https://github.com/robotlocomotion/drake/blob/3a796886bb6cc177431b327a5bd0f74725973f7a/geometry/optimization/graph_of_convex_sets.h#L34)):*
> ```objc
>   bool convex_relaxation{true};
> 
>   /** Round solution after solving convex relaxation. If convex relaxation is
> ```

convex_relaxation
___
*[`geometry/optimization/graph_of_convex_sets.h` line 36 at r1](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NAkYwMu8inyIAkfix3F:-NAkYwMu8inyIAkfix3G:bcxs5lp) ([raw file](https://github.com/robotlocomotion/drake/blob/3a796886bb6cc177431b327a5bd0f74725973f7a/geometry/optimization/graph_of_convex_sets.h#L36)):*
> ```objc
>   /** Round solution after solving convex relaxation. If convex relaxation is
>   false, this option is ignored. */
>   bool round_solution{false};
> ```

this one is more of a decision.  i'm tempted to make it true, though?  @tobiamarcucci , wdyt?
___
*[`geometry/optimization/graph_of_convex_sets.cc` line 607 at r1](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NAkZKh141v14ykjyY5G:-NAkZKh27d8_dqoSsTmg:b-rfdkvu) ([raw file](https://github.com/robotlocomotion/drake/blob/3a796886bb6cc177431b327a5bd0f74725973f7a/geometry/optimization/graph_of_convex_sets.cc#L607)):*
> ```cpp
>   }
> 
>   if (options.convex_relaxation && options.round_solution) {
> ```

cite https://arxiv.org/abs/2205.04422 section 4.2?

**Review by mpetersen94 — COMMENTED** [review]
Added unit tests!

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17796#-:-NBZ0iFbACJ1XY7w3xUs:b3jogwn)* status: 3 unresolved discussions, LGTM missing from assignee RussTedrake(platform), needs at least two assigned reviewers (waiting on @RussTedrake and @TobiaMarcucci)

___
*[`geometry/optimization/graph_of_convex_sets.h` line 34 at r1](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NAkYrVm-j0vvL0ET683:-NBJLZ6i9Yvkv0O_onnp:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3a796886bb6cc177431b327a5bd0f74725973f7a/geometry/optimization/graph_of_convex_sets.h#L34)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

convex_relaxation
</blockquote></details>

Done.
___
*[`geometry/optimization/graph_of_convex_sets.h` line 36 at r1](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NAkYwMu8inyIAkfix3F:-NBYhMZwEVZ99WhQQXxZ:b-ccwb7c) ([raw file](https://github.com/robotlocomotion/drake/blob/3a796886bb6cc177431b327a5bd0f74725973f7a/geometry/optimization/graph_of_convex_sets.h#L36)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

this one is more of a decision.  i'm tempted to make it true, though?  @tobiamarcucci , wdyt?
</blockquote></details>

I'm leaning towards false because it changes the expected output of the solve without deprecation.
___
*[`geometry/optimization/graph_of_convex_sets.cc` line 607 at r1](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NAkZKh141v14ykjyY5G:-NBJM1nf-VkgwaECu3aZ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3a796886bb6cc177431b327a5bd0f74725973f7a/geometry/optimization/graph_of_convex_sets.cc#L607)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

cite https://arxiv.org/abs/2205.04422 section 4.2?
</blockquote></details>

Done.

**Review by RussTedrake — COMMENTED** [review]
> Reviewed 5 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17796)* status: 5 unresolved discussions, LGTM missing from assignee RussTedrake(platform), needs at least two assigned reviewers (waiting on @mpetersen94)

___
*[`geometry/optimization/graph_of_convex_sets.h` line 41 at r2](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NBczvkTDY--euCaD12r:-NBczvkTDY--euCaD12s:b9cnx3f) ([raw file](https://github.com/robotlocomotion/drake/blob/de5f0ddfbccab841723b03fb178206bbed071999/geometry/optimization/graph_of_convex_sets.h#L41)):*
> ```objc
>   /** Round solution after solving convex relaxation. If convex_relaxation is
>   false, this option is ignored. */
>   bool round_solution{false};
> ```

i found myself wondering if this is redundant... does max_rounded_paths = 0 mean the same thing?
___
*[`geometry/optimization/graph_of_convex_sets.h` line 43 at r2](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NBczlc_C0k31ia9XMzs:-NBczlc_C0k31ia9XMzt:b-mwmcon) ([raw file](https://github.com/robotlocomotion/drake/blob/de5f0ddfbccab841723b03fb178206bbed071999/geometry/optimization/graph_of_convex_sets.h#L43)):*
> ```objc
>   bool round_solution{false};
> 
>   /** Maximum number of unique paths to search for during random rounding. If
> ```

perhaps "to return" would be more clear than "to search for", since you try up to max_rounding trials?
___
*[`geometry/optimization/graph_of_convex_sets.h` line 44 at r2](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NBczQd36GYPlMyEwPxY:-NBczQd36GYPlMyEwPxZ:b8v8mlv) ([raw file](https://github.com/robotlocomotion/drake/blob/de5f0ddfbccab841723b03fb178206bbed071999/geometry/optimization/graph_of_convex_sets.h#L44)):*
> ```objc
> 
>   /** Maximum number of unique paths to search for during random rounding. If
>   convex_relaxation is false, this option is ignored. */
> ```

if convex_relaxation is false or round_solution is false... (here and below)
___
*[`geometry/optimization/graph_of_convex_sets.cc` line 867 at r2](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NBd0GeR8GW67gGaXjxe:-NBd0GeR8GW67gGaXjxf:b-7pc97t) ([raw file](https://github.com/robotlocomotion/drake/blob/de5f0ddfbccab841723b03fb178206bbed071999/geometry/optimization/graph_of_convex_sets.cc#L867)):*
> ```cpp
>   }
> 
>   // Implements the rounding scheme put forth in Section 4.2 of
> ```

this seems to be relatively independent of the code above.  should we consider making it a separate class method?  and then the parameters that you've added to the options struct could instead be passed as an argument, along with the MathematicalProgramResult?
___
*[`geometry/optimization/graph_of_convex_sets.cc` line 888 at r2](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NBd3fxc5wRuBeSxcT61:-NBd3fxc5wRuBeSxcT62:bim3lx6) ([raw file](https://github.com/robotlocomotion/drake/blob/de5f0ddfbccab841723b03fb178206bbed071999/geometry/optimization/graph_of_convex_sets.cc#L888)):*
> ```cpp
>       num_trials++;
> 
>       // Find path
> ```

can you say a little more here?  something like "Find a path by traversing the graph depth first with edges taken with probabilities proportional to their flows"?
___
*[`geometry/optimization/graph_of_convex_sets.cc` line 925 at r2](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NBd1Vsd75lCK72hcIDM:-NBd1Vsd75lCK72hcIDN:b-b3o0av) ([raw file](https://github.com/robotlocomotion/drake/blob/de5f0ddfbccab841723b03fb178206bbed071999/geometry/optimization/graph_of_convex_sets.cc#L925)):*
> ```cpp
> 
>       if (std::find(paths.begin(), paths.end(), new_path) != paths.end()) {
>         continue;
> ```

fyi -- i thought for a bit about whether we could do this in the inner loop (e.g. avoid taking the same path to the end twice and then rejection sampling it by keeping yet another list of dirty or candidate vertices/edges) but i guess it's not very easy.  
___
*[`geometry/optimization/graph_of_convex_sets.cc` line 929 at r2](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NBd53u6-ZB3fkOZMObA:-NBd53u6-ZB3fkOZMObB:bin4ia4) ([raw file](https://github.com/robotlocomotion/drake/blob/de5f0ddfbccab841723b03fb178206bbed071999/geometry/optimization/graph_of_convex_sets.cc#L929)):*
> ```cpp
>       paths.push_back(new_path);
> 
>       // Optimize path
> ```

ah.  i see.  this is why you are inside the main solve function.. because you're building on the same mathematical program.  i guess it's not so nice to put things in a different function, afterall.

I'll leave my other comments so that you can see that thought process, but have marked them as "informing" instead of blocking.
___
*[`geometry/optimization/graph_of_convex_sets.cc` line 973 at r2](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NBd1FWs3kDBKDuPmqm0:-NBd1FWs3kDBKDuPmqm1:b-aajw0v) ([raw file](https://github.com/robotlocomotion/drake/blob/de5f0ddfbccab841723b03fb178206bbed071999/geometry/optimization/graph_of_convex_sets.cc#L973)):*
> ```cpp
>       }
>     }
>     result = best_result;
> ```

i would have thought you'd want to return all the rounded paths?  and all the results?  (again, it could work better if RoundShortestPath is a separate method?)
___
*[`geometry/optimization/test/graph_of_convex_sets_test.cc` line 1246 at r2](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NBd6-L_0s7tiE91BEGV:-NBd6-La39Jb_n5ZfNTn:b-lsp9wh) ([raw file](https://github.com/robotlocomotion/drake/blob/de5f0ddfbccab841723b03fb178206bbed071999/geometry/optimization/test/graph_of_convex_sets_test.cc#L1246)):*
> ```cpp
> 
>   spp.AddEdge(*p3, *p1);
>   spp.AddEdge(*p4, *p2);
> ```

this seems an awfully complicated graph to test the basic rounding (especially with cartesian power sets, etc).  

isn't there a simple, almost symmetric graph that would capture it better?
if you do want to keep it this way, can you document why this is the right test?  (e.g. this is a known example from ___ where rounding was shown to be valuable...)

File: geometry/optimization/graph_of_convex_sets.h [hunk 1]
```diff
@@ -36,6 +36,26 @@ struct GraphOfConvexSetsOptions {
   reduction in computation by reducing the size of the optimization solved. */
   bool preprocessing{true};
 
+  /** Round solution after solving convex relaxation. If convex_relaxation is
+  false, this option is ignored. */
+  bool round_solution{false};
+
+  /** Maximum number of unique paths to search for during random rounding. If
```
Comment (TobiaMarcucci) [inline geometry/optimization/graph_of_convex_sets.h:43]: Minor wording: I'd say distinct instead of unique

**Review by TobiaMarcucci — APPROVED** [review]

**Comment by TobiaMarcucci** [comment]
One other high-level comment, mostly to keep track of our experience with this rounding. While we agree that a random rounding is the right thing (since the deterministic version gets tricked when there are multiple optimal solutions), we haven't really understood if rounding forward is better than rounding backward. My sensation is that they are basically equivalent, i.e., they might perform differently from time to time, but neither of the two has a theoretical advantage over the other.

For what concerns the code, we might (maybe in the future) support both the forward and the backward rounding. In that case I feel that the code between the two methods should be substantially the same. And the decision of which method to adopt should be a flag that dictates the order of the bookkeeping in the while loop.

**Review by mpetersen94 — COMMENTED** [review]
I like this thought process. Since we know forward works and we don't think backwards will be better, I'm going to move forward with just forward rounding for this PR. Later we can compare rounding directions.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17796#-:-NBmTgqm7QkX4BpyBYLs:b7dqrdc)* status: 5 unresolved discussions, LGTM missing from assignee RussTedrake(platform), needs at least two assigned reviewers (waiting on @RussTedrake)

___
*[`geometry/optimization/graph_of_convex_sets.h` line 41 at r2](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NBczvkTDY--euCaD12r:-NBmJBwj2X8SxbdBEjlu:bggfldq) ([raw file](https://github.com/robotlocomotion/drake/blob/de5f0ddfbccab841723b03fb178206bbed071999/geometry/optimization/graph_of_convex_sets.h#L41)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

i found myself wondering if this is redundant... does max_rounded_paths = 0 mean the same thing?
</blockquote></details>

Removing it does prevent a corner case failure where round_solution is true but max_rounded_paths <=0 in which case an empty result is returned. Done.
___
*[`geometry/optimization/graph_of_convex_sets.h` line 43 at r2](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NAjnnR4iX-LXcfnV1IS-r2-43:-NBmJBCX9pRQ1XlQ8b3D:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/de5f0ddfbccab841723b03fb178206bbed071999/geometry/optimization/graph_of_convex_sets.h#L43)):*
<details><summary><i>Previously, TobiaMarcucci (Tobia Marcucci) wrote…</i></summary><blockquote>

Minor wording: I'd say distinct instead of unique
</blockquote></details>

Done.
___
*[`geometry/optimization/graph_of_convex_sets.h` line 43 at r2](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NBczlc_C0k31ia9XMzs:-NBmK5xu5thHEExhUkWK:bylgw4f) ([raw file](https://github.com/robotlocomotion/drake/blob/de5f0ddfbccab841723b03fb178206bbed071999/geometry/optimization/graph_of_convex_sets.h#L43)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

perhaps "to return" would be more clear than "to search for", since you try up to max_rounding trials?
</blockquote></details>

Since we only return the lowest cost path I went with some different wording. PTAL.
___
*[`geometry/optimization/graph_of_convex_sets.h` line 44 at r2](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NBczQd36GYPlMyEwPxY:-NBmKOgo6sXcR2o9O3OO:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/de5f0ddfbccab841723b03fb178206bbed071999/geometry/optimization/graph_of_convex_sets.h#L44)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

if convex_relaxation is false or round_solution is false... (here and below)
</blockquote></details>

Done.
___
*[`geometry/optimization/graph_of_convex_sets.cc` line 867 at r2](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NBd0GeR8GW67gGaXjxe:-NBsNCXPCJcd0XIFnDTx:b-iejvfs) ([raw file](https://github.com/robotlocomotion/drake/blob/de5f0ddfbccab841723b03fb178206bbed071999/geometry/optimization/graph_of_convex_sets.cc#L867)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

this seems to be relatively independent of the code above.  should we consider making it a separate class method?  and then the parameters that you've added to the options struct could instead be passed as an argument, along with the MathematicalProgramResult?
</blockquote></details>

I played with this a while. If I try to spin it out as a private method that `SolveShortestPath` calls, I have to pass a lot of data to that method. If I spin it out into a method the user calls separately, then under the hood it will either modify the graph and then call `SolveShortestPath` on the modified graph or I have to duplicate the optimization setup from `SolveShortestPath`. I'm not sure the current code structure is the best, but I liked it better than the other structures I could think of.
___
*[`geometry/optimization/graph_of_convex_sets.cc` line 888 at r2](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NBd3fxc5wRuBeSxcT61:-NBmUpVk8QB5BPhB3pM_:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/de5f0ddfbccab841723b03fb178206bbed071999/geometry/optimization/graph_of_convex_sets.cc#L888)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

can you say a little more here?  something like "Find a path by traversing the graph depth first with edges taken with probabilities proportional to their flows"?
</blockquote></details>

Done.
___
*[`geometry/optimization/graph_of_convex_sets.cc` line 973 at r2](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NBd1FWs3kDBKDuPmqm0:-NBsO73yBaCVRhJMnD8E:bz60pz8) ([raw file](https://github.com/robotlocomotion/drake/blob/de5f0ddfbccab841723b03fb178206bbed071999/geometry/optimization/graph_of_convex_sets.cc#L973)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

i would have thought you'd want to return all the rounded paths?  and all the results?  (again, it could work better if RoundShortestPath is a separate method?)
</blockquote></details>

Generally only the best path is useful. Maybe a future use case could benefit from multiple returned paths, but I don't know it.
___
*[`geometry/optimization/test/graph_of_convex_sets_test.cc` line 1246 at r2](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NBd6-L_0s7tiE91BEGV:-NBmTXmIE9vC4tTVUwke:by2anhy) ([raw file](https://github.com/robotlocomotion/drake/blob/de5f0ddfbccab841723b03fb178206bbed071999/geometry/optimization/test/graph_of_convex_sets_test.cc#L1246)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

this seems an awfully complicated graph to test the basic rounding (especially with cartesian power sets, etc).  

isn't there a simple, almost symmetric graph that would capture it better?
if you do want to keep it this way, can you document why this is the right test?  (e.g. this is a known example from ___ where rounding was shown to be valuable...)
</blockquote></details>

Simplified it a bit and added a comment showing the graph. Any simpler and the relaxation solves it completely. PTAL

**Review by RussTedrake — COMMENTED** [review]
:lgtm:
+@jwnimmer-tri for platform review, please.

> Reviewed 6 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17796#-:-NC070HW0s963xmVyvzU:b-el6yf3)* status: 1 unresolved discussion, LGTM missing from assignee jwnimmer-tri(platform) (waiting on @jwnimmer-tri and @mpetersen94)

___
*[`geometry/optimization/graph_of_convex_sets.h` line 43 at r2](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NBczlc_C0k31ia9XMzs:-NC06IALBiNgyDFVORUD:b-z7od2m) ([raw file](https://github.com/robotlocomotion/drake/blob/de5f0ddfbccab841723b03fb178206bbed071999/geometry/optimization/graph_of_convex_sets.h#L43)):*
<details><summary><i>Previously, mpetersen94 (Mark Petersen) wrote…</i></summary><blockquote>

Since we only return the lowest cost path I went with some different wording. PTAL.
</blockquote></details>

minor tweak: how about "during random rounding; only the lowest cost path is returned."
___
*[`geometry/optimization/graph_of_convex_sets.cc` line 867 at r2](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NBd0GeR8GW67gGaXjxe:-NC06cgG6jfBbSmFjNsF:btzckrb) ([raw file](https://github.com/robotlocomotion/drake/blob/de5f0ddfbccab841723b03fb178206bbed071999/geometry/optimization/graph_of_convex_sets.cc#L867)):*
<details><summary><i>Previously, mpetersen94 (Mark Petersen) wrote…</i></summary><blockquote>

I played with this a while. If I try to spin it out as a private method that `SolveShortestPath` calls, I have to pass a lot of data to that method. If I spin it out into a method the user calls separately, then under the hood it will either modify the graph and then call `SolveShortestPath` on the modified graph or I have to duplicate the optimization setup from `SolveShortestPath`. I'm not sure the current code structure is the best, but I liked it better than the other structures I could think of.
</blockquote></details>

I agree. (I came to the same conclusion)
___
*[`geometry/optimization/test/graph_of_convex_sets_test.cc` line 1228 at r3](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NC06p3-5oD2hHrxEA46:-NC06p3-5oD2hHrxEA47:bbor32g) ([raw file](https://github.com/robotlocomotion/drake/blob/460187f655b13e809605150d6a224370d7f4577f/geometry/optimization/test/graph_of_convex_sets_test.cc#L1228)):*
> ```cpp
>     └───────►│ p2 │◄───►│ p4 │────►│ target │
>              └────┘     └────┘     └────────┘
> ```

btw -- love it!

**Review by mpetersen94 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17796)* status: 2 unresolved discussions (waiting on @jwnimmer-tri and @RussTedrake)

___
*[`geometry/optimization/graph_of_convex_sets.h` line 43 at r2](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NBczlc_C0k31ia9XMzs:-NC1WiSz27NpIJtxERFO:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/de5f0ddfbccab841723b03fb178206bbed071999/geometry/optimization/graph_of_convex_sets.h#L43)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

minor tweak: how about "during random rounding; only the lowest cost path is returned."
</blockquote></details>

Done.
___
*[`geometry/optimization/graph_of_convex_sets.h` line 44 at r3](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NC1Hb6R9pG8vzjsezKX:-NC1Wwj09DW5Nmxye0rt:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/460187f655b13e809605150d6a224370d7f4577f/geometry/optimization/graph_of_convex_sets.h#L44)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

https://drake.mit.edu/styleguide/cppguide.html#Integer_Types

To my understanding, the use of an unsigned integer here is not consistent with our guide.

If we anticipate values in excess of 2**31 then we can use `int64_t` here; otherwise, just `int`.

Possibly the implementation will need to gain a `DRAKE_THROW_UNLESS(max_rounding_trials >= 0);` due to this change.
</blockquote></details>

Done.
___
*[`geometry/optimization/graph_of_convex_sets.cc` line 896 at r3](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NC1JTfcDyqh4hW99B_B:-NC1XATg4Bj0FPqKY1LV:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/460187f655b13e809605150d6a224370d7f4577f/geometry/optimization/graph_of_convex_sets.cc#L896)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

nit https://drake.mit.edu/styleguide/cppguide.html#Preincrement_and_Predecrement
</blockquote></details>

Done.
___
*[`geometry/optimization/graph_of_convex_sets.cc` line 914 at r3](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NC1MIxy3bAQXnkq0BiX:-NC1mUuc1TdUJ3aZtQpI:buw6kth) ([raw file](https://github.com/robotlocomotion/drake/blob/460187f655b13e809605150d6a224370d7f4577f/geometry/optimization/graph_of_convex_sets.cc#L914)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This particular if-case (when edges is empty) is apparently never run by the unit test.  Is there any way to cover it with a test case?
</blockquote></details>

I put that in to cover the corner case of hitting a dead end in depth first search but given the constraints on shortest path, there should not be any flow into a dead end node after solving the relaxation. I can't come up with a case where this will be triggered but I feel wrong removing it.
___
*[`geometry/optimization/graph_of_convex_sets.cc` line 954 at r3](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NC1KHc47UhbRDSzlYUj:-NC1XsVE4V8BPSLfXegJ:b-szcx1f) ([raw file](https://github.com/robotlocomotion/drake/blob/460187f655b13e809605150d6a224370d7f4577f/geometry/optimization/graph_of_convex_sets.cc#L954)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

nit I think an alternative spelling for these will probably compile okay and might be more readable?
</blockquote></details>

Done here and below.
___
*[`geometry/optimization/graph_of_convex_sets.cc` line 971 at r3](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NC1MUGM7EbhHS_QxnqI:-NC1Z69ZEbHKWKeld9uz:bt4ox15) ([raw file](https://github.com/robotlocomotion/drake/blob/460187f655b13e809605150d6a224370d7f4577f/geometry/optimization/graph_of_convex_sets.cc#L971)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

nit This little if-else stanza would be a candidate to de-duplicate, via a little helper function atop this file:

```c++
void Solve(
    const MathematicalProgram& prog,
    const GraphOfConvexSetsOptions& options,
    MathematicalProgramResult* result);
```

It's not quite to the point of being a strict brittleness defect (there's only two copies so far) but possibly worth doing anyway for clarity.
</blockquote></details>

Done. Thanks for the suggestion!
___
*[`geometry/optimization/graph_of_convex_sets.cc` line 973 at r3](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NC1KYei58Zqta7OlvKA:-NC1YBvv8RGSBH-AQsOI:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/460187f655b13e809605150d6a224370d7f4577f/geometry/optimization/graph_of_convex_sets.cc#L973)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

nit Missing punctuation.
</blockquote></details>

Done.
___
*[`geometry/optimization/test/graph_of_convex_sets_test.cc` line 1292 at r3](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NC1Lk9D8Jm5iyJpMr1p:-NC1YOmwDngqyTKzJiux:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/460187f655b13e809605150d6a224370d7f4577f/geometry/optimization/test/graph_of_convex_sets_test.cc#L1292)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

nit
</blockquote></details>

Done.

**Review by mpetersen94 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17796)* status: 2 unresolved discussions (waiting on @jwnimmer-tri and @RussTedrake)

___
*[`geometry/optimization/graph_of_convex_sets.cc` line 914 at r3](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NC1MIxy3bAQXnkq0BiX:-NC1qeF5AL4T9y7ZViRv:b-syh4dq) ([raw file](https://github.com/robotlocomotion/drake/blob/460187f655b13e809605150d6a224370d7f4577f/geometry/optimization/graph_of_convex_sets.cc#L914)):*
<details><summary><i>Previously, mpetersen94 (Mark Petersen) wrote…</i></summary><blockquote>

I put that in to cover the corner case of hitting a dead end in depth first search but given the constraints on shortest path, there should not be any flow into a dead end node after solving the relaxation. I can't come up with a case where this will be triggered but I feel wrong removing it.
</blockquote></details>

Actually what about this fix? I removed the `candidate_edges.size() == 0` code path but added a DRAKE_DEMAND to ensure the code is going down the other path. The note will make it easy to add back in the `candidate_edges.size() == 0` code path if we ever find an example where that is triggered.

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 5 of 5 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17796)* status: 2 unresolved discussions (waiting on @mpetersen94 and @RussTedrake)

___
*[`geometry/optimization/graph_of_convex_sets.cc` line 914 at r3](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NC1MIxy3bAQXnkq0BiX:-NC1rHW39aSrB0t9kmnd:bb6ywmx) ([raw file](https://github.com/robotlocomotion/drake/blob/460187f655b13e809605150d6a224370d7f4577f/geometry/optimization/graph_of_convex_sets.cc#L914)):*
<details><summary><i>Previously, mpetersen94 (Mark Petersen) wrote…</i></summary><blockquote>

Actually what about this fix? I removed the `candidate_edges.size() == 0` code path but added a DRAKE_DEMAND to ensure the code is going down the other path. The note will make it easy to add back in the `candidate_edges.size() == 0` code path if we ever find an example where that is triggered.
</blockquote></details>

LGTM
___
*[`geometry/optimization/graph_of_convex_sets.cc` line 617 at r4](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NC1qtNSA10XEjuFbfX8:-NC1qtNSA10XEjuFbfX9:b4luf9g) ([raw file](https://github.com/robotlocomotion/drake/blob/4e4a9fb7e765055353ab1810e1fa6b4882c0a33d/geometry/optimization/graph_of_convex_sets.cc#L617)):*
> ```cpp
> }
> 
> MathematicalProgramResult Solve(const MathematicalProgram& prog,
> ```

nit The helper function should have internal linkage (https://drake.mit.edu/styleguide/cppguide.html#Internal_Linkage)



_Suggestion:_
```cpp
namespace {
MathematicalProgramResult Solve(const MathematicalProgram& prog,
                                const GraphOfConvexSetsOptions& options) {
  MathematicalProgramResult result;
  if (options.solver) {
    options.solver->Solve(prog, {}, options.solver_options, &result);
  } else {
    result = solvers::Solve(prog, {}, options.solver_options);
  }
  return result;
}
}  // namespace anonymous
```

**Review by mpetersen94 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17796)* status: 1 unresolved discussion (waiting on @jwnimmer-tri and @RussTedrake)

___
*[`geometry/optimization/graph_of_convex_sets.cc` line 617 at r4](https://reviewable.io/reviews/robotlocomotion/drake/17796#-NC1qtNSA10XEjuFbfX8:-NC22rScAdczWApObXUC:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/4e4a9fb7e765055353ab1810e1fa6b4882c0a33d/geometry/optimization/graph_of_convex_sets.cc#L617)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

nit The helper function should have internal linkage (https://drake.mit.edu/styleguide/cppguide.html#Internal_Linkage)
</blockquote></details>

Done.

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17796)* status: 1 unresolved discussion (waiting on @RussTedrake)

**Review by jwnimmer-tri — COMMENTED** [review]
> Dismissed @RussTedrake from a discussion.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17796)* status: :shipit: complete! all discussions resolved, LGTM from assignees jwnimmer-tri(platform),RussTedrake(platform) (waiting on @mpetersen94)
